File: C++/TwoPointer/lengthOfLongestSubstring.cpp
/*
Solution:
- 利用双指针思想，采用滑动窗口；时间复杂度O(n).
- 利用字典来存储字符出现过的位置，设置左右指针i,j分别指向子串的左右边界；
- 当j所指向字符不在i-j之间出现，j右移,如果出现这个字符，那么i右移，这个新的子串从重复字符出现位置的后面一个位置开始；并判断maxLen与j-i的大小；
- 关于字符是否出现用dict[s[j]] >= i来判断，dict[s[j]]表示j所指向字符最近一次出现的位置；
*/
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.length() == 0){
            return s.length();
        }
        vector<int> dict(256, -1);
        dict[s[0]] = 0;
        int maxLen = 1, i = 0, j = 1;
        while(i <= j && j < s.length()){
            if(dict[s[j]] >= i){
                maxLen = max(maxLen, j-i);    
                i = dict[s[j]] + 1;
            }
            else{
                dict[s[j]] = j;
                j++;
            }
            if(i == j){
                dict[s[i]] = i;
                j++;
            }
        }
        maxLen = max(maxLen, j-i);
        return maxLen;
    }
};

File: Leetcode高频题/lengthOfLongestSubstring.py
'''
@lsy 2019.10.21

Solution:
双指针。

使用字典dic来存储目前已经出现过的所有字符的最远的距离。
指针i，j。移动指针j，
- 若当前字符未在dic中出现，加入dic
- 若当前字符在dic中出现，判断当前字符之前出现的最远位置，如果在i,j之间，说明出现重复字符，需要将i指针移动到dic[s[j]] + 1
  更新当前字符出现的最远位置。

时间复杂度O(n)，空间复杂度O(n)
'''
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int: 
        if len(s) < 2:
            return len(s)
        
        dic, cnt = {}, 0
        i, j = 0, 0
        while j < len(s):
            c = s[j]
            if c in dic.keys() and i <= dic[c] < j:
                i = dic[c] + 1
                
            dic[c] = j
            j += 1
            cnt = max(cnt, j-i)
            
        return cnt
                
        
        